### PR TITLE
Add bindFirst and curried methods for Poly

### PR DESCRIPTION
**Pull request by Atry**

### CODE REVIEW

**Comment by Atry** [comment]
I thought this PR should not break the backward-compatibility, however the build log error with the following error messages. Why?
```
[info] core: found 2 potential binary incompatibilities while checking against com.chuusai:shapeless_2.11:2.3.0  (filtered 7)
[error]  * method curried()shapeless.PolyDefns#Curried in trait shapeless.Poly is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("shapeless.Poly.curried")
[error]  * method bindFirst(java.lang.Object)shapeless.PolyDefns#BindFirst in trait shapeless.Poly is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("shapeless.Poly.bindFirst")
java.lang.RuntimeException: core: Binary compatibility check failed!
	at scala.sys.package$.error(package.scala:27)
	at com.typesafe.tools.mima.plugin.SbtMima$.reportModuleErrors(SbtMima.scala:110)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6$$anonfun$apply$4.apply(MimaPlugin.scala:57)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6$$anonfun$apply$4.apply(MimaPlugin.scala:48)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:109)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6.apply(MimaPlugin.scala:48)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$6.apply(MimaPlugin.scala:42)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (coreJVM/*:mimaReportBinaryIssues) core: Binary compatibility check failed!
[error] Total time: 3 s, completed Apr 21, 2017 12:27:45 PM
/home/travis/.travis/job_stages: line 153:  2259 Terminated              travis_jigger $! $timeout $cmd
```

**Comment by Atry** [comment]
@milessabin Why did the code coverage change?

**Comment by milessabin** [comment]
This looks very useful ... thanks!

A few things before I merge. Do these work as expected as arguments to a `Mapper`? If not, then I suspect they would if the instances were moved to the corresponding `CaseN` companions. Either way, would you mind adding some tests of this case.

Also some stylistic things. shapeless intentionally doesn't use access modifiers (other than `private[this]` where it's interaction with variance can be useful). Also please could you adjust your use of whitespace to match the rest of the project (ie. two space indents with aggressive line breaking).

**Comment by Atry** [comment]
Yes, `Mapper` is my use case, too.

However, there is a limitation: 

The newly created `Poly` from `bindFirst` or `curried` must be visible at the call site, which is impossible for some complicated case.

**Comment by milessabin** [comment]
Still waiting on tests and the stylistic changes requested.

**Comment by joroKr21** [comment]
Rebased to remove whitespace from diff and some tests with `map` and `foldLeft`.
@milessabin regarding stylistic changes I think if we want a consistent style we should setup scalafmt.

**Comment by Atry** [comment]
Thank you for picking up this PR @joroKr21

File: core/src/main/scala/shapeless/poly.scala [hunk 1]
```diff
@@ -113,6 +113,59 @@ object PolyDefns extends Cases {
...
+        finalCall: Case.Aux[F, Head :: Tail, Result0]
+    ): Case.Aux[BF, Tail, Result0] = new Case[BF, Tail] {
+      type Result = Result0
+      val value: Tail => Result = { tail: Tail =>
+        finalCall.value(witnessBF.value.head :: tail)
+      }
+    }
+  }
+
+  final case class Curried[F, ParameterAccumulator <: HList](parameters: ParameterAccumulator) extends Poly1
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/poly.scala]: @Atry do you remember what the `ParameterAccumulator` type parameter is for?
It seems unused to me = always `HNil`

Comment (Atry) [reply core/src/main/scala/shapeless/poly.scala]:
```
    object p extends Poly3 {
      implicit def x = at[Int, Double, String] { (i, d, s) =>
        s"$i, $d, $s"
      }
    }

    val c = Poly.curried(p)
    val c1 = c(1)
    val c2 = c1(42.5)
```

Try printing `c2`.

Comment (joroKr21) [reply core/src/main/scala/shapeless/poly.scala]: Ah ok that makes sense :+1:

**Pull request merged**